File: src/common/vcs/git.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Vcs_utils

type error_status = Vcs_utils.error_status =
  | Not_installed of { path: string }
  | Errored of string

let git ?cwd args = exec ?cwd "git" args

(** [merge_base a b] returns the hash of the common ancestor of [a] and [b].

    See https://git-scm.com/docs/git-merge-base for caveats. In particular,
    if [a] and [b] have multiple common ancestors, which one git returns
    is undefined (we don't pass [--all]). We also don't pass [--fork-point],
    so we may return "upstream" changes if history was rewritten. *)
let merge_base ?cwd a b =
  match%lwt git ?cwd ["merge-base"; a; b] with
  | Ok stdout ->
    let result =
      try Ok (String.sub stdout 0 40) with
      | Invalid_argument _ -> Error (Errored "Malformed merge-base")
    in
    Lwt.return result
  | Error _ as err -> Lwt.return err

let files_changed_since ?cwd hash =
  match%lwt git ?cwd ["diff"; "-z"; "--name-only"; hash] with
  | Ok stdout -> Lwt.return (Ok (split_null_terminated_lines stdout))
  | Error _ as err -> Lwt.return err

let files_changed_since_mergebase_with ?cwd commit =
  match%lwt merge_base ?cwd commit "HEAD" with
  | Error _ as err -> Lwt.return err
  | Ok mergebase ->
    (match%lwt files_changed_since ?cwd mergebase with
    | Ok changes -> Lwt.return (Ok (mergebase, changes))
    | Error _ as err -> Lwt.return err)

File: rust_port/crates/flow_common_vcs/src/git.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

pub use crate::vcs_utils::ErrorStatus;
use crate::vcs_utils::exec;
use crate::vcs_utils::split_null_terminated_lines;

async fn git(cwd: Option<&str>, args: &[&str]) -> Result<String, ErrorStatus> {
    exec(None, cwd, "git", args).await
}

pub async fn merge_base(cwd: Option<&str>, a: &str, b: &str) -> Result<String, ErrorStatus> {
    match git(cwd, &["merge-base", a, b]).await {
        Ok(stdout) => {
            if stdout.len() >= 40 {
                Ok(stdout[..40].to_string())
            } else {
                Err(ErrorStatus::Errored("Malformed merge-base".to_string()))
            }
        }
        Err(err) => Err(err),
    }
}

pub async fn merge_base_and_timestamp(
    cwd: Option<&str>,
    a: &str,
    b: &str,
) -> Result<(String, i64), ErrorStatus> {
    match merge_base(cwd, a, b).await {
        Err(err) => Err(err),
        Ok(hash) => match git(cwd, &["log", "--format=%ct", "-n1", &hash]).await {
            Ok(stdout) => match stdout.trim().parse::<i64>() {
                Ok(timestamp) => Ok((hash, timestamp)),
                Err(_) => Err(ErrorStatus::Errored(format!(
                    "Failed to convert timestamp {:?}",
                    stdout
                ))),
            },
            Err(err) => Err(err),
        },
    }
}

pub async fn files_changed_since(
    cwd: Option<&str>,
    hash: &str,
) -> Result<Vec<String>, ErrorStatus> {
    match git(cwd, &["diff", "-z", "--name-only", hash]).await {
        Ok(stdout) => Ok(split_null_terminated_lines(&stdout)
            .into_iter()
            .map(|s| s.to_string())
            .collect()),
        Err(err) => Err(err),
    }
}

pub async fn files_changed_since_mergebase_with(
    cwd: Option<&str>,
    commit: &str,
) -> Result<(String, Vec<String>), ErrorStatus> {
    match merge_base(cwd, commit, "HEAD").await {
        Err(err) => Err(err),
        Ok(mergebase) => match files_changed_since(cwd, &mergebase).await {
            Ok(changes) => Ok((mergebase, changes)),
            Err(err) => Err(err),
        },
    }
}
